tools/reviewer/web/src/app/shared/services/localserver.service.ts

```diff
@@ -89,28 +90,30 @@ export class LocalserverService {
 
   getDiffFiles(id: number, workspace: string): Observable<File[]> {
     return new Observable(observer => {
-      this.getBranchInfo(id, workspace).subscribe(diffFilesResponse => {
-        let files: File[] = [];
-        for (const commit of diffFilesResponse.getCommitList()) {
-          // Remove duplicates
-          let newfiles: File[] = commit.getFileList();
-          newfiles = newfiles.filter(newfile => {
-            for (const file of files) {
-              if (newfile.getFilename() === file.getFilename()) {
-                return false;
-              }
-            }
-            return true;
-          });
-
-          files = files.concat(newfiles);
-        }
-
-        observer.next(files);
+      this.getBranchInfo(id, workspace).subscribe(branchInfo => {
+        observer.next(this.getFilesFromBranchInfo(branchInfo));
       });
     });
   }
 
+  getFilesFromBranchInfo(branchInfo: BranchInfo): File[] {
+    const fileDictionary = new Dictionary<File>();
+    for (const commit of branchInfo.getCommitList()) {
+      this.addWithReplace(fileDictionary, commit.getFileList());
+    }
+    this.addWithReplace(fileDictionary, branchInfo.getUncommittedFileList());
+
+    return fileDictionary.values;
+  }
+
+  addWithReplace(fileDictionary: Dictionary<File>, newfiles: File[]): void {
+    for (const file of newfiles) {
+      // If file with the filename is already exist, it will be replaced by
+      // the new one.
+      fileDictionary.add(file.getFilename(), file);
```

Comment:
I still don't understand why `DiffFilesResponse` returns an array of `BranchInfo` instead of just `BranchInfo`. Therefore web app always uses first element only.
https://github.com/google/startup-os/blob/master/tools/reviewer/web/src/app/shared/services/localserver.service.ts#L51

Reply:
About commits. Of course there might be different versions of a file on a several commits.
But we don't need all the versions, we need only 1.
We use file list in 2 places:
1) Displaying all files in a diff page. We don't care about commits here. We need filenames only.
2) Searching specific file by filename from file list. At this point we uses file with most newest commit, and it's exactly what the file list contains.

Reply:
Diff deals with multiple repos :)
A workspace can have multiple repos (all repos under `base/head`), and Diff can potentially have changes in all of them.

Reply:
Aside from that, as long as you're aware of this, that's fine.

Reply:
> Diff deals with multiple repos :)

In that case how are we going to know which file should be used in `file-change` page?
`https://startupos-5f279.firebaseapp.com/diff/65/tools/aa/AaModule.java`
Maybe the path should look like that:
`https://startupos-5f279.firebaseapp.com/diff/65/startup-os/tools/aa/AaModule.java`
(with repo)

Reply:
Also, filenames could be kept like:
`startup-os/tools/aa/AaModule.java`
To be unique.